src/components/views/dialogs/DeactivateAccountDialog.js

```diff
@@ -0,0 +1,139 @@
+
+        this.state = {
+            confirmButtonEnabled: false,
+            busy: false,
+            errStr: null,
+        };
+    }
+
+    _onPasswordFieldChange(ev) {
+        this.setState({
+            confirmButtonEnabled: Boolean(ev.target.value),
+        });
+    }
+
+    _onOk() {
+        // This assumes that the HS requires password UI auth
+        // for this endpoint. In reality it could be any UI auth.
+        this.setState({busy: true});
+        MatrixClientPeg.get().deactivateAccount({
+            type: 'm.login.password',
+            user: MatrixClientPeg.get().credentials.userId,
+            password: this._passwordField.value,
+        }).done(() => {
+            Lifecycle.onLoggedOut();
+            this.props.onFinished(false);
+        }, (err) => {
+            let errStr = 'Unknown error';
+            // https://matrix.org/jira/browse/SYN-744
+            if (err.httpStatus == 401 || err.httpStatus == 403) {
+                errStr = 'Incorrect password';
+                Velocity(this._passwordField, "callout.shake", 300);
+            }
+            this.setState({
+                busy: false,
+                errStr: errStr,
+            });
+        });
+    }
+
+    _onCancel() {
+        this.props.onFinished(false);
+    }
+
+    render() {
+        const Loader = sdk.getComponent("elements.Spinner");
+        let passwordBoxClass = '';
+
+        let error = null;
+        if (this.state.errStr) {
+            error = <div className="error">
+                {this.state.err_str}
+            </div>
+            passwordBoxClass = 'error';
+        }
+
+        // I would use a spinner here but that causes the button to be rendered
+        // slightly higher than the other one
+        const okLabel = this.state.busy ? <Loader /> : 'Deactivate Account';
+        const okEnabled = this.state.confirmButtonEnabled && !this.state.busy;
+
+        let cancelButton = null;
+        if (!this.state.busy) {
+            cancelButton = <button onClick={this._onCancel} autoFocus={true}>
+                Cancel
+            </button>
+        }
+
+        return (
+            <div className="mx_DeactivateAccountDialog">
+                <div className="mx_Dialog_title danger">
+                    Deactivate Account
+                </div>
+                <div className="mx_Dialog_content">
+                    This will make your account permanently unusable. You will not be able to re-register the same user ID.<br />
+                    <br />
+                    This action is irreversible.<br />
+                    <br />
+                    To continue, please enter your password.<br />
+                    <br />
+                    Password:<br />
+                    <input
+                        type="password"
+                        onChange={this._onPasswordFieldChange}
+                        ref={(e) => {this._passwordField = e;}}
```

Comment:
(see https://facebook.github.io/react/docs/more-about-refs.html#the-ref-callback-attribute): 

> also note that when writing refs with inline function expressions as in the examples here, React sees a different function object each time so on every update, ref will be called with null immediately before it's called with the component instance.